### ISSUE TITLE
replace call to functions from the string module in web

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5004]
|<img alt="nueces's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| nueces reported|
|-|-|
|Trac ID|trac#5004|
|Type|enhancement|
|Created|2011-03-27 17:43:41Z|
|Branch|https://github.com/twisted/twisted/tree/twisted-web-remove-text-5004-2|

To make the port to python 3k more easy, I'm replacing all the calls to functions from the string module, for their equivalent from the str type object or calling directly to the built-in method of the string variable.

Attachments:

* [web.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/3596a9b816b562e13eb3206d54d67d5ee9452898.patch) (5834 bytes) - added by nueces on 2011-03-27 17:43:54Z - patch
* [replace-string-module-web-5004.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/3ca13ab5f131b3acd4cee11ecca47c6883b74c48.patch) (3025 bytes) - added by multani on 2013-04-28 09:08:56Z - 
* [replace-string-module-web-5004.patch-2](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/d963a72d7a5a0f8e00d32341a539a5f09f564a78.patch-2) (8377 bytes) - added by multani on 2013-05-08 10:48:30Z - 
* [replace-string-module-web-5004-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/22818fc6775cfd780abe03877c50e16b86342579.patch) (8379 bytes) - added by multani on 2013-05-08 19:07:57Z - 
* [replace-string-module-web-5004-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/600cc6a391fd86cc4c7e379da04b1286b6e8fe72.patch) (8385 bytes) - added by multani on 2013-05-08 19:09:56Z - 
* [replace-string-module-web-5004-5.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/e3ef2e0ba0559608a064678da3f867bcf3ad04e6.patch) (8364 bytes) - added by multani on 2013-06-19 12:23:24Z - 
* [replace-string-module-web-5004-6.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/a7bd8b988a4d1760d124e519bc16890655070368.patch) (9933 bytes) - added by multani on 2013-07-07 11:12:26Z - 
* [replace-string-module-web-5004-7.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/38b968a4b50c69c83afcaf0fa3ca186a3724001a.patch) (10768 bytes) - added by multani on 2013-07-31 18:00:17Z - 
* [replace-string-module-web-5004-8.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/3ac/3aca064ac13d759f44ddd4e9cc148d8b193e994b/3c7660d6dd54c9a67bf074b14bb1c87b234d2f9e.patch) (2145 bytes) - added by multani on 2013-08-30 15:47:04Z - On top of branches/twisted-web-remove-text-5004 [39084]

<details><summary>Searchable metadata</summary>

```
trac-id__5004 5004
type__enhancement enhancement
reporter__nueces nueces
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_twisted_web_remove_text_5004_2 branches/twisted-web-remove-text-5004-2
branch_author__tomprince__rwall tomprince, rwall
status__closed closed
resolution__fixed fixed
component__web web
keywords__py3k py3k
time__1301247821000000 1301247821000000
changetime__1384965703000000 1384965703000000
version__None None
owner__rwall rwall
cc__jknight cc__thijs cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to nueces|
|-|-|

Thanks for working on this.  The code changed here (sadly :/) isn't executed by the Twisted Web unit tests.  Can you add unit tests that verify this code continues to work properly after this change?  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

My previous review was based on some bogus results from a coverage reporting tool.  Actually most of the changes are to code executed by the test suite.  However, the change in `getSession` is not tested.  Can you add a test just for that?  Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_8"></a>|@multani set owner to @multani<br>@multani set status to `assigned`|
|-|-|

I updated the patch so that all remaining call to functions of the string module has been replace by the equivalent string methods.

If there are still some needs to improve test coverage, I'd gladly do so, what is the process to get the coverage with the Python 3 current scripts?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

Thanks for your patch.

 1. Running pyflakes on `twisted.web.server` shows `types` is also unused and can be removed:
```
twisted/web/server.py:13: 'string' imported but unused
twisted/web/server.py:14: 'types' imported but unused
```
 1. only line 91 in `twisted.web.twcgi` doesn't have coverage, can you add a test for this line?
 1. pyflakes reports an unused import in `twisted.web.util` that can be removed:
```
twisted/web/util.py:18: 'FilePath' imported but unused
```
 1. `htmlIndent` on line 202 doesn't have test coverage

ps. also un-assign yourself from the ticket next time you add the `review` keyword, thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

Replying to [multani](#note_8):
> what is the process to get the coverage with the Python 3 current scripts?

I'm not sure if this documented somewhere, other than the buildbot slave that runs it, but to check coverage I use something like this that creates a html coverage report:

```
#!sh
#!/bin/bash

rm -rf _trial_temp
rm -rf twisted-coverage
coverage run --omit /usr/*,_trial_temp/* --branch ./bin/trial --reporter=bwverbose twisted
coverage html -d twisted-coverage --omit /usr/*,_trial_temp/*,Divmod/trunk/* -i
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

> > what is the process to get the coverage with the Python 3 current scripts?
> I'm not sure if this documented somewhere, other than the buildbot slave that runs it, but to check coverage I use something like this that creates a html coverage report

Note this doesn't generate a coverage report for the code running under Python 3.  I'm not sure if coverage.py itself has been ported to Python 3 (I guess it probably has but I haven't checked).  The single-codebase approach to porting means that the Python 3 coverage is *usually* pretty similar to the Python 2 coverage though.  If you have full Python 2 coverage then you usually don't need to worry about what the Python 3 coverage looks like.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_12"></a>|@multani removed owner<br>@multani set status to `new`|
|-|-|

Replying to [thijs](#note_9):
> Thanks for your patch.
> 
>  1. Running pyflakes on `twisted.web.server` shows `types` is also unused and can be removed:
> ```
> twisted/web/server.py:13: 'string' imported but unused
> twisted/web/server.py:14: 'types' imported but unused
> ```

I removed types.

>  1. only line 91 in `twisted.web.twcgi` doesn't have coverage, can you add a test for this line?

I added a test for this line. I had to change a few things in ```DummyRequest``` since it wasn't implementing the whole ```Request``` interface as needed by ```CGIScript```

>  1. pyflakes reports an unused import in `twisted.web.util` that can be removed:
> ```
> twisted/web/util.py:18: 'FilePath' imported but unused
> ```

This is removed.

>  1. `htmlIndent` on line 202 doesn't have test coverage

I added tests for this as well.

> ps. also un-assign yourself from the ticket next time you add the `review` keyword, thanks.

Sorry about that!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra set owner to @multani|
|-|-|

Thanks for your patch. Some minor things;

 1. It appears you used `git diff` which produces a different patch format compared to `svn diff`
 1. 2 blank lines between methods and tests
 1. tests should be named like `test_escapeHtml` instead of `test_escape_html`

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_14"></a>|@multani commented|
|-|-|

Thanks for the review.

Replying to [thijs](#note_13):
> Thanks for your patch. Some minor things;
> 
>  1. It appears you used `git diff` which produces a different patch format compared to `svn diff`

As I did for the first patch. If you are referring to Trac not able to show up the patch correctly against the repository, I would say it's because of the ending extension which is not ```.patch```. I followed http://twistedmatrix.com/trac/wiki/GitMirror in both cases.

>  1. 2 blank lines between methods and tests

Done.

>  1. tests should be named like `test_escapeHtml` instead of `test_escape_html`

Done.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_17"></a>|@lvh set owner to @multani|
|-|-|

Hi multani! Thanks for working on this ticket.

Looks mostly good! Two questions from the diff (I get line numbers from https://twistedmatrix.com/trac/attachment/ticket/5004/replace-string-module-web-5004-4.patch):

twisted/web/twcgi.py, line 53 in the new file, and later 58, 59: why doesn't this just do _reactor=reactor? It seems reactor behaves as the default value here.

twisted/web/twcgi.py, line 104 in the new file uses str.find; maybe it's a good idea to turn this into str.index and replace the if/else with try/except since str.find is pretty terrible (it works now, because it explicitly checks for the (dumb) return value of str.find, but I think str.index would be an improvement).

Otherwise, LGTM.

Thanks again for working on this
lvh

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

I like the idiom:

```
def foo(name=None):
    if name is None:
        name = default
    ...
```

In particular, it makes it easier for callers to invoke the default behavior.  It's easier and safer to pass None than to hard-code the actual default value of a parameter at the call-site (or write two code-paths, one which passes a value for a parameter if you have a good value to pass, one that doesn't if you don't).

In any case, the coding standard doesn't call for one or the other, so I think this part of the patch is fine.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_19"></a>|@multani removed owner|
|-|-|

Replying to [lvh](#note_17):
> Hi multani! Thanks for working on this ticket.

Thanks for the review!

Beside your reivew, I also rebased the patch on the latest trunk.

> Looks mostly good! Two questions from the diff (I get line numbers from https://twistedmatrix.com/trac/attachment/ticket/5004/replace-string-module-web-5004-4.patch):
> 
> twisted/web/twcgi.py, line 53 in the new file, and later 58, 59: why doesn't this just do _reactor=reactor? It seems reactor behaves as the default value here.

I tend to agree with exarkun on this. I'm not very opiniated though, so I'm still willing to change it if it helps to get this merged.

> twisted/web/twcgi.py, line 104 in the new file uses str.find; maybe it's a good idea to turn this into str.index and replace the if/else with try/except since str.find is pretty terrible (it works now, because it explicitly checks for the (dumb) return value of str.find, but I think str.index would be an improvement).

Yeah, ```.index``` is a nice improvement, so this is now done.

(FYI, it's also available as a Git branch at https://github.com/multani/twisted/tree/replace-string-module-web-5004)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_20"></a>|@tomprince commented|
|-|-|

(In [38927]) Branching to twisted-web-remove-text-5004.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_21"></a>|@tomprince commented|
|-|-|

(In [38928]) Apply replace-string-module-web-5004-5.patch from multani.

Refs: [#5004](https://github.com/twisted/twisted/issues/5004)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince set owner to @multani|
|-|-|

1.`twcgi` and the reactor:
   1. Persuant to [this](https://twistedmatrix.com/pipermail/twisted-python/2013-June/027132.html) mailing list thread, please make both the argument and attribute public.
   2. `FilteredScript` should also use the specified reactor (this will need a test)
   3. Since the default value is the only place the global reactor will be used, please import it conditionally if required. This means that importing this module won't (directly) import the reactor (thus installing the default).
 2. I started a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twisted-web-remove-text-5004).

Other than that, this look good. Please resubmit, with a patch against the branch.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_23"></a>|@multani commented|
|-|-|

Replying to [tom.prince](#note_22):
>  1. `twcgi` and the reactor:
>    1. Persuant to [this](https://twistedmatrix.com/pipermail/twisted-python/2013-June/027132.html) mailing list thread, please make both the argument and attribute public.

I understand for the argument, but should I really make the attribute public as well?

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_24"></a>|@tomprince commented|
|-|-|

Replying to [multani](#note_23):
> >  1. `twcgi` and the reactor:
> >    1. Persuant to [this](https://twistedmatrix.com/pipermail/twisted-python/2013-June/027132.html) mailing list thread, please make both the argument and attribute public.
> 
> I understand for the argument, but should I really make the attribute public as well?

I think I had a reasonable motivation for suggesting that, when I wrote that. But reflecting on it, I can't think of a good reason for it. Furthermore, it is easy to change it at a later date to make it public, but not the reverse. So erring on the side of making it private seems reasonable.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_25"></a>|@multani commented|
|-|-|

Replying to [tom.prince](#note_22):
>  1. `twcgi` and the reactor:
>    1. Persuant to [this](https://twistedmatrix.com/pipermail/twisted-python/2013-June/027132.html) mailing list thread, please make both the argument and attribute public.

Done.

>    2. `FilteredScript` should also use the specified reactor (this will need a test)

Done. I had to change slightly the test's tear down, which supposed the custom `startServer` method was called (which I don't for my tests).

Actually, I guess those custom tests method could go away, since we can now passes a custom reactor to `FilteredScript` instead of starting a real one.

>    3. Since the default value is the only place the global reactor will be used, please import it conditionally if required. This means that importing this module won't (directly) import the reactor (thus installing the default).

Done.

>  2. I started a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twisted-web-remove-text-5004).
> 
> Other than that, this look good. Please resubmit, with a patch against the branch.

I rebased the patch against latest trunk.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince commented|
|-|-|

(In [39081]) Apply replace-string-module-web-5004-6.patch from multani.

Refs: [#5004](https://github.com/twisted/twisted/issues/5004)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_27"></a>|@tomprince commented|
|-|-|

Thanks for your continued work on this.
 1. Please add docstrings to the methods added to `DummyRequest`.
   - When looking at them, I noticed that `getHost` returns a transport. But `IRequest.getHost` returns an `IAddress`, not a transport. Please change it to return an `IAddress` (probably an `IPv4Address`).
 2. I fixed a few twistedchecker errors: some whitespace fixes, and `s/_fake_reactor/fakeReactor/`.
 3. I reverted the change back of `index` back to `find`.

Please resubmit for review after addressing 1 with a patch against the branch.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_28"></a>|@multani commented|
|-|-|

Sorry it took so long for just a few docstrings...

Replying to [tom.prince](#note_27):
> Thanks for your continued work on this.
>  1. Please add docstrings to the methods added to `DummyRequest`.
>    - When looking at them, I noticed that `getHost` returns a transport. But `IRequest.getHost` returns an `IAddress`, not a transport. Please change it to return an `IAddress` (probably an `IPv4Address`).


I added the docstrings and fixed the returned value of `getHost`.
I wasn't so much inspired for the docstrings, so they are merely based on what `twisted.web.http.Request` presents.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_30"></a>|@hawkowl set owner to @multani|
|-|-|

Hi multani! Thanks for resubmitting.

1. Should "L{htmlIndent} transparently process input with no special cases inside." have "processes" instead of "process"? (changing this will require reflowing the docstring)
2. Should "L{htmlIndent} replaces tab characters by unbreakable spaces." have "with" rather than "by"?
3. Maybe "Retrieve all the values of the request headers as a dictionary." should use "get" to fall in line with the rest of the docstrings, as well as the function name?

Otherwise, I don't see anything obviously wrong with it, and I don't see any code problems that the above haven't fixed - please fix these and resubmit.

**Comment by twisted-trac** [comment]
|[<img alt="multani's avatar" src="https://avatars.githubusercontent.com/u/65311?s=50" width="50" height="50">](https://github.com/multani)<a name="note_31"></a>|@multani commented|
|-|-|

Hi hawkowl, thanks for the review.

I'm proposing a new patch on top of the branch proposed by Tom in comment:26 which fixes your two first points. I let "Retrieve" instead of "Get" since this was also the term used by the previous ```getHeader`` method.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_32"></a>|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

Reviewing...

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_33"></a>|@wallrj commented|
|-|-|

(In [40816]) Apply replace-string-module-web-5004-8.patch from multani. Refs [#5004](https://github.com/twisted/twisted/issues/5004)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_34"></a>|@wallrj set status to `new`|
|-|-|

Thanks multani,

This looks pretty good to me. Just a few nit picks which I've listed below.

Notes:
 * Applied the latest replace-string-module-web-5004-8.patch
 * Merges cleanly
 * Forced build: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twisted-web-remove-text-5004&num_builds=10
   * Some spurious pyflakes warnings.

Points:
 1. Missing newsfile -- I'll add one.
    1. https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles
 2. Some twistedchecker warnings
    1. https://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1536/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
 3. There is some missing coverage adjacent to some code you changed, but none
    of the other reviewers noted it, so I think we can ignore it.
    1. https://buildbot.twistedmatrix.com/builds/twisted-coverage.py/twisted-coverage.py-r40816/twisted_web_twcgi.html#n107
 4. [branches/twisted-web-remove-text-5004/twisted/web/util.py](branches/twisted-web-remove-text-5004/twisted/web/util.py)
    1. Non-standard line wrapping on 201, 202
       1. See https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#auto5
 5. [branches/twisted-web-remove-text-5004/twisted/web/test/requesthelper.py](branches/twisted-web-remove-text-5004/twisted/web/test/requesthelper.py)
    1. getAllHeaders: Missing @return documentation
    2. getClient: Unnecessary "pass" statement
 6. [branches/twisted-web-remove-text-5004/twisted/web/test/test_cgi.py](branches/twisted-web-remove-text-5004/twisted/web/test/test_cgi.py)
    1. test_useReactorArgument: Missing docstrings for FakeReactor
    2. def test_pathInfo: the assertIn test for the presence of the key looks
       unnecessary, since you're testing the value of that key later anyway.

I think I'll go ahead and fix those and merge if the build results come back
green.

Thanks again to everyone who's worked on this ticket.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_35"></a>|@wallrj commented|
|-|-|

(In [40817]) Branching to 'twisted-web-remove-text-5004-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_36"></a>|@wallrj commented|
|-|-|

Addressed code review comments:
 * log:branches/twisted-web-remove-text-5004-2

Build results:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/twisted-web-remove-text-5004-2

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_37"></a>|@wallrj set status to `closed`|
|-|-|

(In [40826]) Merge twisted-web-remove-text-5004-2

Author: nueces, multani
Reviewers: exarkun, thijs, lvh, tom.prince, hawkowl, rwall
Fixes: [#5004](https://github.com/twisted/twisted/issues/5004)

Replace all use of the string module in twisted.web, with the equivalent string
methods.
